src/functional.jl

```diff
@@ -0,0 +1,111 @@
+module FunctionalNullableOperations
+
+# extends Base with operations that treat nullables as collections
+
+importall Base
+import Base: promote_op, LinearFast
+
+# conceptually, nullable types can be considered infinite-order tensor powers of
+# finite-dimensional vector spaces — we attempt to support most operations on
+# arrays except the linear algebra related ones; the infinite-dimensional nature
+# makes subtyping AbstractArray a little dangerous, which explains why
+# functionality is reimplemented instead of subtyping AbstractArray
+
+# size   – not implemented since an infinite-dimensional tuple would be strange
+# length – one or zero
+# endof  – same as length
+length(u::Nullable) = u.isnull ? 0 : 1
+endof(u::Nullable)  = length(u)
+
+# indexing is either without index, or with 1 as index
+# generalized linear indexing is not supported
+# setindex! not supported because Nullable is immutable
+linearindexing{T}(::Nullable{T}) = LinearFast()
+function getindex(u::Nullable)
+    @boundscheck u.isnull && throw(NullException())
+    u.value
+end
+function getindex(u::Nullable, i::Integer)
+    @boundscheck u.isnull | (i ≠ one(i)) && throw(BoundsError(i, u))
+    u.value
+end
+
+# iteration protocol
+start(u::Nullable) = 1
+next(u::Nullable, i::Integer) = u.value, 0
+done(u::Nullable, i::Integer) = u.isnull || i == 0
+
+# next we have reimplementations of some higher-order functions
+filter{T}(p, u::Nullable{T}) = u.isnull ? u : p(u.value) ? u : Nullable{T}()
```

Comment:
I actually can't eliminate the branch in general, since `x.value` isn't defined for certain types (`Union{}` in particular). `NullableArrays` uses an `isbits` test for this. Once I optimize for performance, I'll try out a similar trick.